src/main_context_futures.rs

```diff
@@ -0,0 +1,392 @@
+// Copyright 2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use std::mem;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::thread;
+
+use futures_core;
+use futures_core::executor::{Executor, SpawnError};
+use futures_core::task::{LocalMap, UnsafeWake, Waker};
+use futures_core::{Async, Future, Never};
+use futures_executor;
+use futures_util::future::FutureExt;
+
+use MainContext;
+use MainLoop;
+use Source;
+use Priority;
+use ::source::CallbackGuard;
+
+use ::translate::{from_glib_none, from_glib_full, mut_override, ToGlib};
+use ffi as glib_ffi;
+
+// We can't use an enum here because we want to store this in an atomic variable
+const INIT: usize = 0;
+const NOT_READY: usize = 1;
+const READY: usize = 2;
+const DONE: usize = 3;
+
+#[repr(C)]
+struct TaskSource {
+    source: glib_ffi::GSource,
+    future: Option<(Box<Future<Item = (), Error = Never>>, Box<LocalMap>)>,
+    thread: Option<thread::ThreadId>,
+    state: AtomicUsize,
+}
+
+unsafe impl UnsafeWake for TaskSource {
+    unsafe fn clone_raw(&self) -> Waker {
+        Waker::new(glib_ffi::g_source_ref(mut_override(&self.source)) as *const TaskSource)
+    }
+
+    unsafe fn drop_raw(&self) {
+        glib_ffi::g_source_unref(mut_override(&self.source));
+    }
+
+    unsafe fn wake(&self) {
+        if self.state
+            .compare_and_swap(NOT_READY, READY, Ordering::SeqCst) == READY
+        {
+            glib_ffi::g_source_set_ready_time(mut_override(&self.source), 0);
+        }
+    }
+}
+
+unsafe extern "C" fn prepare(
+    source: *mut glib_ffi::GSource,
+    timeout: *mut i32,
+) -> glib_ffi::gboolean {
+    let _guard = CallbackGuard::new();
+
+    let source = &mut *(source as *mut TaskSource);
+
+    *timeout = -1;
+
+    let mut cur = source
+        .state
+        .compare_and_swap(INIT, NOT_READY, Ordering::SeqCst);
+    if cur == INIT {
+        // XXX: This is not actually correct, we should not dispatch the
+        // GSource here already but we need to know its current status so
+        // that if it is not ready yet something can register to the waker
+        if let Async::Ready(_) = source.poll() {
```

Comment:
`_` can be only `()` ?

Reply:
Yes, all tasks/futures that are spawned on the executor must result either `()` or `Never` (i.e. they also must be infallible)

Reply:
Then maybe use `Async::Ready(())` in if lets ?

Reply:
Done